lib/go-tc/deliveryservices.go

```diff
@@ -192,18 +194,416 @@ type DeliveryServiceFieldsV31 struct {
 // DeliveryServiceV40 is a Delivery Service as it appears in version 4.0 of the
 // Traffic Ops API.
 type DeliveryServiceV40 struct {
```

Comment:
I did that because I copy/pasted it in from the "DS Active Flag" changes, where that was actually necessary because it's changing the type of something that was in the lowest-level struct in that "inheritance" tree. I can change it to use nesting, but as soon as we make a breaking change it'll need to do this anyway - for example this PR is still changing a couple of things: the DS match list in API < v4 is a pointer to a slice, which is an unnecessary double-null-check headache because slices are already "nullable", and the `lastUpdated` field is currently in the special ATC format but this changes it to use RFC3339 by switching from `TimeNoMod` to `time.Time`. Those changes seemed too small to warrant their own PR to me, so I left them in, but I'll take them out if you want. I don't think switching time formats should be done across all objects in a single PR, because for things that use the CRUDer that'll mean either doubling the amount of code per endpoint to allow backward compatibility, or rewriting them to not use the CRUDer. I think. Depending on the "flavor" of CRUDer they use.

Reply:
I see. It seems like the DS active flag changes are on the back-burner still, so if we make this change now but have to cut 6.0.x before that is implemented, we'll end up with yet another mostly-duplicated struct.

Reply:
I was hoping #4996 would get added to the 6.x milestone, but it isn't so far. But still, I'd like to keep the timestamps as RFC3339, even if that would mean duplication.

Unfortunately, I think mostly duplicated structs are just a fact of life with a versioned API in Go; even with the nesting of previous versions we have compatibility problems such that clients of the library are forbidden from constructing the objects with direct field initialization like
```go
f := FooVX{
    Property: "value,
}
```
Using nesting saves code changes for non-breaking changes, but makes it harder to make breaking changes.

If you want me to take out all the breaking structure changes I absolutely will, to be clear, since they aren't really necessary to accomplish the PR's goal. But then I intend to open a PR to re-introduce them, and I just hoped they could be small enough to fit into this, since I was touching the struct anyway.

Reply:
> Unfortunately, I think mostly duplicated structs are just a fact of life with a versioned API in Go

I refuse to accept this statement. We need to be aggressively removing old API versions so that we can collapse the nested structs back into one struct over time.

Can you make the RFC3339 change without de-embedding the struct? Sure, that would change it for prior API versions, but that seems like a worthwhile breakage (if it even does break anything).

Reply:
> Can you make the RFC3339 change without de-embedding the struct? Sure, that would change it for prior API versions, but that seems like a worthwhile breakage (if it even does break anything).

I actually think that might break Traffic Portal and possibly any other web-based client of the API used heavily for Delivery Service interactions that may or may not exist. It's possible it wouldn't - one of those is much easier to test than the other.

>> Unfortunately, I think mostly duplicated structs are just a fact of life with a versioned API in Go

> I refuse to accept this statement.

Well, in truth, it's more about what we're willing to give up. not duplicating things is going to mean that you can write code today that won't compile tomorrow. It's just that I'd rather duplicate things than deal with that, personally.

> We need to be aggressively removing old API versions

Agree. I think we can deprecate v2 and v3 with ATCv6 as soon as APIv1 is actually removed.

Reply:
Ok, there might be another option. If you add a new `LastUpdated time.Time` field with the same JSON tags (and probably db tags for good measure) to `DeliveryServiceV40`, when doing JSON marshalling at V40, it will take precedence over the _embedded_ `LastUpdated`. We just have to make sure to set both fields before marshalling the response.

Reply:
Wait, but if it takes precedence in database reads and JSON unmarshaling, then if I get a `DeliveryServiceV4` from either source and try to access older versions through embedding, then won't we run into
```go
ds := somethingThatReturnsADeliveryServiceV4()
older := ds.DeliveryServiceV31

// Outputs: false
// because older.LastUpdated is unitialized
fmt.Println(older.LastUpdated != nil && ds.LastUpdated == older.LastUpdated.Time)
```
Giving a struct two properties of the same name seems brittle and error-prone, even though it can, apparently, work.

Idk, I think the length of this discussion proves I was wrong to think this change was too small to warrant its own PR, seems like I should probably take it out so it isn't the focus of a PR that wasn't primarily meant to do that.

Reply:
Yeah, that is what I meant by this:
> We just have to make sure to set both fields before marshalling the response.

If the value is set in either place, it needs to be set in both places (basically when reading from the DB).

Anyways, small PRs are best PRs 👍 . I'm not the biggest fan of having two fields with the same name, but it seems less prone to bugs than having entirely duplicated structs and would go away once we manage to remove the older API versions.